#### Импорт необходимых библиотек

In [1]:
import pandas as pd

#### Загрузка данных

In [3]:
bookings = pd.read_csv('2_bookings.csv', sep =';')
bookings.head()

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

In [4]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

### Вопрос 1: Исправить внешний вид колонок датасета (верхний регистр, пробелы между словами)

#### Функция, которая в названиях колонках датасета заменяет пробелы на нижнее подчеркивание и приводит названия к нижнему регистру  

In [7]:
def rename_c(columns):
    new_columns = columns.replace(' ','_').lower()
    return (new_columns)

#### Применяю метод rename, в качестве аргумента параметра columns указываю написанную выше функцию

In [9]:
bookings=bookings.rename(columns=rename_c)
bookings.head()

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

### Вопрос 2: Пользователи из каких стран совершили наибольшее число успешных бронирований?

In [10]:
top5_book=bookings.query("is_canceled==0").country.value_counts()
top5_book.head()

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

#### Ответ: Португалия, Великобритания, Франция, Испания, Германия

### Вопрос 3: На сколько ночей (stays_total_nights) в среднем бронируют отели типа City Hotel? Resort Hotel?


In [11]:
bookings.hotel.unique()
bookings.groupby('hotel', as_index=False).agg({'stays_total_nights':'mean'}).round(2)

hotel  stays_total_nights
0    City Hotel                2.98
1  Resort Hotel                4.32

#### Ответ: City Hotel - 2.98, Resort Hotel - 4.32

### Вопрос 4: Иногда тип номера, присвоенного клиенту (assigned_room_type),отличается от изначально забронированного (reserved_room_type).Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?

In [12]:
bookings.query('assigned_room_type != reserved_room_type')

hotel  is_canceled  lead_time arrival_full_date  \
2       Resort Hotel            0          7        2015-07-01   
12      Resort Hotel            0         68        2015-07-01   
15      Resort Hotel            0         68        2015-07-01   
17      Resort Hotel            0         12        2015-07-01   
18      Resort Hotel            0          0        2015-07-01   
...              ...          ...        ...               ...   
119273    City Hotel            0        213        2017-08-28   
119274    City Hotel            0        213        2017-08-28   
119289    City Hotel            0         25        2017-08-30   
119297    City Hotel            0        332        2017-08-31   
119357    City Hotel            0         47        2017-08-31   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
2                    2015               July                        27   
12                   2015               July                        27   
15                   2015               July                        27   
17                   2015               July                        27   
18                   2015               July                        27   
...                   ...                ...                       ...   
119273               2017             August                        35   
119274               2017             August                        35   
119289               2017             August                        35   
119297               2017             August                        35   
119357               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
2                               1                        0   
12                              1                        0   
15                              1                        0   
17                              1                        0   
18                              1                        0   
...                           ...                      ...   
119273                         28                        1   
119274                         28                        1   
119289                         30                        0   
119297                         31                        0   
119357                         31                        1   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
2                          1  ...       1       0.0       0    BB     GBR   
12                         4  ...       2       0.0       0    BB     USA   
15                         4  ...       2       0.0       0    BB     IRL   
17                         1  ...       2       0.0       0    BB     IRL   
18                         1  ...       2       0.0       0    BB     FRA   
...                      ...  ...     ...       ...     ...   ...     ...   
119273                     3  ...       1       0.0       0    HB     PRT   
119274                     3  ...       1       0.0       0    HB     PRT   
119289                     3  ...       3       0.0       0    BB     ITA   
119297                     2  ...       2       0.0       0    BB     GBR   
119357                     3  ...       1       0.0       0    SC     PRT   

       reserved_room_type assigned_room_type    customer_type  \
2                       A                  C        Transient   
12                      D                  E        Transient   
15                      D                  E        Transient   
17                      A                  E        Transient   
18                      A                  G        Transient   
...                   ...                ...              ...   
119273                  A                  K  Transient-Party   
119274                  A                  K  Transient-Party   
119289                  E                  F        Transient   
119297       

#### Ответ: 14917

### Вопрос 5: На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?

In [13]:
bookings.arrival_date_year.unique()
bookings.groupby('arrival_date_year').arrival_date_month.agg(pd.Series.mode)

arrival_date_year
2015    September
2016      October
2017          May
Name: arrival_date_month, dtype: object

#### Ответ: в 2016 - Октябрь, в 2017 - Май

### Вопрос 6: Сгруппируйте данные по годам, а затем проверьте, на какой месяц (arrival_date_month) бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017? 

In [14]:
bookings.query('hotel=="City Hotel"').groupby(['arrival_date_year', 'arrival_date_month']).is_canceled.sum()

arrival_date_year  arrival_date_month
2015               August                1232
                   December               668
                   July                   939
                   November               301
                   October               1321
                   September             1543
2016               April                 1539
                   August                1247
                   December              1072
                   February               930
                   January                438
                   July                  1043
                   June                  1720
                   March                 1108
                   May                   1436
                   November              1360
                   October               1947
                   September             1567
2017               April                 1926
                   August                1123
                   February               

#### Ответ: в 2015 - на Сентябрь, в 2016 - на Октябрь, в 2017 - на Май

### Вопрос 7: Посмотрите на числовые характеристики трёх колонок: adults, children и babies. Какая из них имеет наибольшее среднее значение?

In [16]:
bookings[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

#### Ответ: adults - 1.856403

### Вопрос 8: Создайте колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?

In [17]:
bookings['total_kids'] = bookings.children + bookings.babies
bookings.groupby('hotel').total_kids.mean().round(2)

hotel
City Hotel      0.10
Resort Hotel    0.14
Name: total_kids, dtype: float64

#### Ответ: для отелей типа Resort Hotel

### Вопрос 9:  Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе(Churn Rate).
####  Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Далее проверьте, среди какой группы пользователей показатель оттока выше.

In [19]:
def has_kids(x):
    if x > 0:
        return True
    else:
        return False

bookings['has_kids'] = bookings.total_kids.apply(has_kids)
bookings['has_kids'] = bookings.total_kids.astype(bool)
bookings[['has_kids', 'is_canceled']].value_counts() / bookings [['has_kids']].value_counts() *100

has_kids  is_canceled
False     0              62.780998
          1              37.219002
True      0              65.049272
          1              34.950728
dtype: float64

#### Ответ: показатель оттока выше среди клиентов, которые при бронировании указывали наличие ребенка - 37.219002